src/audio/smart_amp/smart_amp_maxim_dsm.c

```diff
@@ -0,0 +1,593 @@
+		goto err;
+	excur[0] = cmdblock[1];
+	excur[1] = cmdblock[2];
+
+	cmdblock[0] = DSM_SET_CMD_ID(19);
+	retcode = dsm_api_get_params(hspk->dsmhandle, 1, (void *)cmdblock);
+	if (retcode != DSM_API_OK)
+		goto err;
+	temp[0] = cmdblock[1];
+	temp[1] = cmdblock[2];
+
+	cmdblock[0] = DSM_SET_CMD_ID(18);
+	retcode = dsm_api_get_params(hspk->dsmhandle, 1, (void *)cmdblock);
+	if (retcode != DSM_API_OK)
+		goto err;
+	dc_res[0] = cmdblock[1];
+	dc_res[1] = cmdblock[2];
+
+	cmdblock[0] = DSM_SET_CMD_ID(16);
+	retcode = dsm_api_get_params(hspk->dsmhandle, 1, (void *)cmdblock);
+	if (retcode != DSM_API_OK)
+		goto err;
+	fc[0] = cmdblock[1];
+	fc[1] = cmdblock[2];
+
+	cmdblock[0] = DSM_SET_CMD_ID(17);
+	retcode = dsm_api_get_params(hspk->dsmhandle, 1, (void *)cmdblock);
+	if (retcode != DSM_API_OK)
+		goto err;
+	q[0] = cmdblock[1];
+	q[1] = cmdblock[2];
+
+	comp_dbg(dev, "[DSM] excursion(Q27) :%08x, %08x",
+		 excur[0], excur[1]);
+	comp_dbg(dev,
+		 "[DSM] temp(Q19) :%08x, %08x, dc_res(Q27) :%08x, %08x",
+		 temp[0], temp[1], dc_res[0], dc_res[1]);
+	comp_dbg(dev,
+		 "[DSM] fc(Q9) :%08x, %08x, q(Q29) :%08x, %08x",
+		 fc[0], fc[1], q[0], q[1]);
+
+	return 0;
+err:
+	comp_err(dev, "[DSM] Get adaptive parameters. error:%d", retcode);
+	return retcode;
+}
+#endif
+
+static inline void *_memmove(void *dest, const void *src, size_t count)
+{
+	char *tmp;
+	const char *s;
+
+	if (dest <= src) {
+		tmp = dest;
+		s = src;
+		while (count--)
+			*tmp++ = *s++;
+	} else {
+		tmp = dest;
+		tmp += count;
+		s = src;
+		s += count;
+		while (count--)
+			*--tmp = *--s;
+	}
+
+	return dest;
+}
+
+static void maxim_dsm_ff_proc(struct smart_amp_mod_struct_t *hspk,
+			      struct comp_dev *dev, void *in, void *out,
+			      int nsamples, int szsample)
+{
+	union smart_amp_buf buf, buf_out;
+	int16_t *input = (int16_t *)hspk->buf.input;
+	int16_t *output = (int16_t *)hspk->buf.output;
+	int *w_ptr = &hspk->buf.ff.avail;
+	int *r_ptr = &hspk->buf.ff_out.avail;
+	bool is_16bit = szsample == 2 ? true : false;
+	int remain;
+	int x;
+
+	buf.buf16 = (int16_t *)hspk->buf.ff.buf;
+	buf.buf32 = (int32_t *)hspk->buf.ff.buf;
+	buf_out.buf16 = (int16_t *)hspk->buf.ff_out.buf;
+	buf_out.buf32 = (int32_t *)hspk->buf.ff_out.buf;
+
+	/* Current pointer(w_ptr) + number of input frames(nsamples)
+	 * must be smaller than buffer size limit
+	 */
+	if (*w_ptr + nsamples <= DSM_FF_BUF_DB_SZ) {
+		if (is_16bit)
+			memcpy_s(&buf.buf16[*w_ptr], nsamples * szsample,
+				 in, nsamples * szsample);
+		else
+			memcpy_s(&buf.buf32[*w_ptr], nsamples * szsample,
+				 in, nsamples * szsample);
+		*w_ptr += nsamples;
+	} else {
+		comp_err(dev,
+			 "[DSM] Feed Forward buffer overflow. (w_ptr : %d + %d > %d)",
+			 *w_ptr, nsamples, DSM_FF_BUF_DB_SZ);
+		return;
+	}
+
+	/* Run DSM Feedforward process if the buffer is ready */
+	if (*w_ptr >= DSM_FF_BUF_SZ) {
+		for (x = 0; x < DSM_FRM_SZ; x++) {
+			/* Buffer ordering for DSM : LRLR... -> LL...RR... */
+			if (is_16bit) {
```

Comment:
This is being checked on every loop iteration. best to
```c
if (is_16bit) {
    for (loop) {16bit loop }
else
    for(loop) {32 bit loop}
```

Reply:
Agreed. Shall fix this.